exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java

```diff
@@ -470,43 +500,73 @@ public void eval() {
+            java.time.Period period;
+
+            if (from.isAfter(to) && duration.compareTo(java.time.Duration.ZERO) > 0) {
+                // negative period and positive duration
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate().plusDays(1));
+                duration = duration.minusDays(1);
+            } else if (from.isBefore(to) && duration.compareTo(java.time.Duration.ZERO) < 0) {
+                // positive period and negative duration
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate().minusDays(1));
+                duration = duration.plusDays(1);
+            } else {
+                period = java.time.Period.between(from.toLocalDate(), to.toLocalDate());
+            }
+
+            out.months = (int) period.toTotalMonths();
+            out.days = period.getDays();
+            out.milliseconds = (int) duration.toMillis();
         }
     }
 
     @FunctionTemplate(name = "age", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls = NullHandling.NULL_IF_NULL)
     public static class AgeDate2Function implements DrillSimpleFunc {
         @Param DateHolder right;
-        @Workspace long queryStartDate;
+        @Workspace java.time.OffsetDateTime queryStartDate;
         @Output IntervalHolder out;
         @Inject ContextInformation contextInfo;
 
         @Override
         public void setup() {
             int timeZoneIndex = contextInfo.getRootFragmentTimeZone();
-            org.joda.time.DateTimeZone timeZone = org.joda.time.DateTimeZone.forID(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
-            org.joda.time.DateTime now = new org.joda.time.DateTime(contextInfo.getQueryStartTime(), timeZone);
-            queryStartDate = (new org.joda.time.DateMidnight(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), timeZone)).getMillis();
+            java.time.ZoneId zoneId = java.time.ZoneId.of(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
+            java.time.ZonedDateTime dt = java.time.Instant.ofEpochMilli(contextInfo.getQueryStartTime()).atZone(zoneId);
+            queryStartDate = java.time.OffsetDateTime.of(dt.toLocalDate(), java.time.LocalTime.MIDNIGHT, java.time.ZoneOffset.UTC);
         }
 
         @Override
         public void eval() {
-            long diff = queryStartDate - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            java.time.OffsetDateTime dt = java.time.Instant.ofEpochMilli(right.value).atOffset(java.time.ZoneOffset.UTC);
```

Comment:
The above has a bug due to a horrible Drill bug.

Drill's "timestamp" is relative to 1/1/970 local time, not UTC. The `right.value` is a number of milliseconds since 1970-01-01'T'00:00:00 **local time**. This is true even if the underlying data source has dates in the (currently unsupported) UTC time zone, unless the (virtual) machine running Drill runs in the UTC time zone.

This is horrible, wrong in the era of big data, but it is what it is. It was explained to me, as "it is what customer's expect because that's how Oracle does it", and Oracle does it that way because the original DATE, TIME and DATETIME had no implied time zone, they were "the time wherever you happen to be" because they were defined in the late 1970s.

In terms of Java time, we have to convert the ms arguments into a `LocalTime`. It appears that `Period` will "do the right thing" if given two `LocalTime` arguments.

Fortunately, here, since we are computing an `AGE()`, we get the right answer if we work in either UTC or local time.

Reply:
@paul-rogers 
In such a case, it might be worth checking the implementation of converting a string to a timestamp.
For example, in CastVarCharToTimeStamp, UTC is used when recording the result:
out.value = parsedDateTime.toInstant(java.time.ZoneOffset.UTC).toEpochMilli();

Reply:
@paul-rogers 
Moreover, the DateUtility.parseBest implementation simply discards the time offset information.

Reply:
@oleg-zinovev, as @dzamo has clearly suggested, Drill's date/time code is, to put it kindly, a mess. We mix zone-less DATE and TIME with a local-time-based TIMESTAMP, then we carelessly cast from local time to UTC. Every bit of time-related code does its own hacks as bugs are discovered. Would be great to actually fix this stuff, but all we are usually able to do is muddle though as best we can for the task at hand.

Code which knows it is using UTC generally converts from local time to UTC. Code which knows it is working with local time generally does local time calculations. The problem is the zone-less DATE and TIME. Java Time does provide classes for those, but historically Drill code simply converted them to local time.

Further, historically, local time was represented using classes meant for UTC, further adding to the confusion. Drill timestamps are similar to Unix timestamps, in that they are milliseconds from 1970-01-01, but they use 1970-01-01'T'00:00:00 *local time* as the "epoch", not the Unix epoch.

(The problem is that, in most modern distributed systems, local time is so much of a problem that everything, except Drill, runs in UTC. Sigh...)

Here, as @dzamo suggests, we have to make some assumptions:

* All TIMESTAMP fields are in local time. (If a field was in UTC in the input file, it must have been converted to local time by the corresponding reader.)
* All DATE fields are converted to midnight local time.
* The query start timestamp is converted to local time.
* AGE is the number of calendar days, local time, between the query start time and the data column.

Can we instead do the math in UTC? No, we can't. Your birthday, for example, occurs from midnight to midnight in whatever timezone you happen to be in at that time. If you were born in London, but now live in San Francisco, your birthday does not occur in London time and thus from 4PM the day before to 4PM on your birthday in San Francisco. Said another way, you are a year older on midnight of your birthday in whatever timezone you happen to be in at that moment.

Dates are generally difficult; Drill, because of how vague it has been, simply adds to the difficulty.